pkg/virt-controller/watch/migration.go

```diff
@@ -49,84 +70,161 @@ func (md *MigrationDispatch) Execute(store cache.Store, queue workqueue.RateLimi
+		queue.Forget(key)
+		return
+	}
 
-			numOfPods, migrationPodExists := investigateTargetPodSituation(migration, podList)
-			if numOfPods > 1 && !migrationPodExists {
-				logger.Error().Reason(err).Msg("another migration seems to be in progress, marking Migration as failed")
-				// Another migration is currently going on
-				setMigrationFailed(migrationCopy)
-				return
-			} else if numOfPods == 1 && !migrationPodExists {
-				// We need to start a migration target pod
-				// TODO, this detection is not optimal, it can lead to strange situations
-				err := md.vmService.SetupMigration(migration, vm)
-				if err != nil {
-					logger.Error().Reason(err).Msg("creating a migration target node failed")
-					queue.AddRateLimited(key)
-					return
-				}
-			}
-			logger.Error().Msg("another migration seems to be in progress, marking Migration as failed")
-			migrationCopy.Status.Phase = v1.MigrationInProgress
-			// TODO indicate when this has happened
-			err = md.vmService.UpdateMigration(migrationCopy)
+	// Fetch vm which we want to migrate
+	vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
+	if err != nil {
+		logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
+		queue.AddRateLimited(key)
+		return
+	}
+	if !exists {
+		logger.Info().Msgf("VM with name %s does not exist, marking migration as failed", migration.Spec.Selector.Name)
+		setMigrationFailed(migration)
+		return
+	}
+	if vm.Status.Phase != v1.Running {
+		logger.Error().Msgf("VM with name %s is in state %s, no migration possible. Marking migration as failed", vm.GetObjectMeta().GetName(), vm.Status.Phase)
+		setMigrationFailed(migration)
+		return
+	}
+
+	if err := mergeConstraints(migration, vm); err != nil {
+		logger.Error().Reason(err).Msg("merging Migration and VM placement constraints failed.")
+		queue.AddRateLimited(key)
+		return
+	}
+	podList, err := md.vmService.GetRunningVMPods(vm)
+	if err != nil {
+		logger.Error().Reason(err).Msg("could not fetch a list of running VM target pods")
+		queue.AddRateLimited(key)
+		return
+	}
+
+	numOfPods, targetPod := investigateTargetPodSituation(migration, podList)
+
+	if targetPod == nil {
+		if numOfPods > 1 {
+			logger.Error().Reason(err).Msg("another migration seems to be in progress, marking Migration as failed")
+			// Another migration is currently going on
+			setMigrationFailed(migration)
+			return
+		} else if numOfPods == 1 {
+			// We need to start a migration target pod
+			// TODO, this detection is not optimal, it can lead to strange situations
+			err := md.vmService.SetupMigration(migration, vm)
 			if err != nil {
-				logger.Error().Reason(err).Msgf("updating migration state failed : %v ", err)
+				logger.Error().Reason(err).Msg("creating a migration target node failed")
 				queue.AddRateLimited(key)
 				return
 			}
 		}
+		setMigrationPhase(migration, v1.MigrationInProgress)
+		return
+	}
+
```

Comment:
Also think about the possibility to safe conditions on a controller object, to have controller internal substates ...